### PR TITLE
Event logger script support: Behavior Context reflection

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
## What does this PR do?

 Adding BehaviorContext support for using the EventLogger API

The primary new functions are `CreateMetricsLogger`, `RecordEvent`
`IsEventLoggerRegistered` and `GetGlobalEventLoggerFactory`.

Several wrapper classes have been added to BehaviorContext to help with
recording event data.
The `EventLoggerId` class is used to specify an ID for an event logger.
In script it can be created as such
```
import azlmbr.metrics

eventLoggerId1 = azlmbr.metrics.CreateFromInt(1)
eventLoggerId2 = azlmbr.metrics.CreateFromString("foo")
```

The `EventPhase` class is used to specify the type of event to record
```
import azlmbr.metrics

durationEventPhase = azlmbr.metrics.EventPhase.DurationBegin
```

The `InstantEventScope` class is used to specify the scope values to an
InstantEvent.
```
import azlmbr.metrics

globalScope = azlmbr.metrics.InstantEventScope.Global
```

The `EventValue` class is used to create a value that can be appended to
the `EventArgs` structure for configuring events
```
import azlmbr.metrics

intValue = azlmbr.emtrics.EventValue(42)
stringValue = azlmbr.metrics.EventValue("Hello world")
```

The `EventArgs` class is the primary class used to configure arguments
to record for an event
```
import azlmbr.metrics

myArgs = azlmbr.metrics.EventArgs()
myArgs.m_name = "Duration Event"
myArgs.m_cat = "Test Category"
myArgs.m_id = "0"
myArgs.AppendArg("Entity Count", 20)
myArgs.AppendArg("Entity Group", "level")

loggerId = azlmbr.metrics.CreateFromString("CustomIdValue")
if azlmbr.metrics.RecordEvent(loggerId, azlmbr.metrics.EventPhase.DurationBegin, myArgs):
    print("Success")

```

For completeness the `CreateMetricsLogger` function can be used as
follows
```
import pathlib
import azlmbr.metrics

loggerId = azlmbr.metrics.CreateFromString("CustomFoo")
outputFilePath = pathlib.Path(azlmbr.paths.projectroot) /
"user/Metrics/MyScriptFoo_metrics.json"
loggerName = "LoggerNameIsFoo"
azlmbr.metrics.CreateMetricsLogger(loggerId, outputFilePath.as_posix(), loggerName)
```

Other changes include updating the BehaviorContext and BehaviorClass to
have methods for querying a BehaviorClass, BehaviorMethod,
BehaviorProperty, property setter, property getter, or BehaviorEBus via
name.

A Scoped wrapper for the BehaviorObject has been added which
automatically destructs and optionally deallocates objects created by
the Behavior Context.

The BehaviorMethod class has had a IsCallable function which returns a
boolean that indicates if function could be invoked with a given set of
input arguments. If it cannot an outcome with the error message
detailing the reason is returned.

## How was this PR tested?

Added C++ UnitTest that verifies the new Behavior Context functions for the EventLogger

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Metrics/EventLoggerReflectUtils.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace AZ
+{
+    class ReflectContext;
+    class BehaviorContext;
+}
+
+namespace AZ::Metrics
+{
+    // Utility namespace contains functions/class to simpilfy calls to record events
```
Comment (lsemp3d) [inline Code/Framework/AzCore/AzCore/Metrics/EventLoggerReflectUtils.h:19]: typo: `simpilfy`

File: Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h [hunk 2]
```diff
@@ -616,6 +620,29 @@ namespace AZ
         static constexpr bool Specialized() { return true; }          \
     };
 
+    /**
+    * Use this macro outside a class to allow it to be identified across modules and serialized (in different contexts).
+    * The expected input is the class and the assigned uuid as a string or an instance of a uuid.
+    * Note that the AZ_TYPE_INFO_SPECIALIZE always has be declared in "namespace AZ".
```
Comment (lsemp3d) [inline Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h:626]:
```suggestion
    * Note that the AZ_TYPE_INFO_SPECIALIZE always has to be declared in "namespace AZ".
```

File: Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h [hunk 3]
```diff
@@ -1264,26 +1291,26 @@ namespace AZ
...
+    AZ_TYPE_INFO_SPECIALIZE(unsigned long, "{5EC2D6F7-6859-400f-9215-C106F5B10E53}");
+    AZ_TYPE_INFO_SPECIALIZE(AZ::u64, "{D6597933-47CD-4fc8-B911-63F3E2B0993A}");
+    AZ_TYPE_INFO_SPECIALIZE(float, "{EA2C3E90-AFBE-44d4-A90D-FAAF79BAF93D}");
+    AZ_TYPE_INFO_SPECIALIZE(double, "{110C4B14-11A8-4e9d-8638-5051013A56AC}");
+    AZ_TYPE_INFO_SPECIALIZE(bool, "{A0CA880C-AFE4-43cb-926C-59AC48496112}");
+    AZ_TYPE_INFO_SPECIALIZE(AZ::Uuid, "{E152C105-A133-4d03-BBF8-3D4B2FBA3E2A}");
+    AZ_TYPE_INFO_SPECIALIZE(void, "{C0F1AFAD-5CB3-450E-B0F5-ADB5D46B0E22}");
+    AZ_TYPE_INFO_SPECIALIZE(Crc32, "{9F4E062E-06A0-46D4-85DF-E0DA96467D3A}");
+    AZ_TYPE_INFO_SPECIALIZE(PlatformID, "{0635D08E-DDD2-48DE-A7AE-73CC563C57C3}");
+    AZ_TYPE_INFO_SPECIALIZE(AZStd::monostate, "{B1E9136B-D77A-4643-BE8E-2ABDA246AE0E}");
```
Comment (lsemp3d) [inline Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h:1313]: What happened here? they were deleted and restored? Also, signed char is commented out, remove?

File: Code/Framework/AzCore/Tests/Metrics/EventLoggerReflectUtilsTests.cpp [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    constexpr const char* EventFieldClassName = "EventField";
+    constexpr const char* EventArrayClassName = "EventArray";
+    constexpr const char* EventObjectClassName = "EventObject";
+
+    TEST_F(EventLoggerReflectUtilsFixture, EventLogger_EventDataTypes_AreAccessibleInScript_Succeeds)
+    {
+        auto classIter = m_behaviorContext->m_classes.find(EventValueClassName);
+        EXPECT_NE(m_behaviorContext->m_classes.end(), classIter);
+
+        classIter = m_behaviorContext->m_classes.find("EventLoggerId");
```
Comment (lsemp3d) [inline Code/Framework/AzCore/Tests/Metrics/EventLoggerReflectUtilsTests.cpp:64]: A few lines above here you `constexpr const char*` define `EventValue`, `EventField` and others. Any reason not to do the same for `EventLoggerId`, `EventArgs`, etc?

**Review by lsemp3d — APPROVED** [review]
Just a few minor things

File: Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h (hunk 3 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/RTTI/TypeInfo.h:1313]: The macro was changed from using `AZ_TYPE_INFO_INTERNAL_SPECIALIZE` to the public `AZ_TYPE_INFO_SPECIALIZE`.

The `signed char` was commented out before and remains commented out. I can remove it.

File: Code/Framework/AzCore/Tests/Metrics/EventLoggerReflectUtilsTests.cpp (hunk 4 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Metrics/EventLoggerReflectUtilsTests.cpp:64]: No reason not to. I can add constants to all of them.

**Review by cgalvan — APPROVED** [review]

**Pull request merged**